### PR TITLE
Support create function CA for compound assignment

### LINKED ISSUE #42170: [Bug]: `Create function` code action not visible inside a statement
### Description

```
function test() {
    string s = "";
    foreach int i in 0...9 {
        foreach int y in 0...9 {
            s += func(); // Create a function code action cannot be found here
        }

    }
}
```

Screenshot 2024-02-16 at 10 53 39 (2)

### Steps to Reproduce

_No response_

### Affected Version(s)

_No response_

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by KavinduZoysa**
## Purpose

Fixes #42170

### CODE REVIEW

**Review by nipunayf — COMMENTED** [review]
Can we add the following two test cases just to cover the possible combinations of the compound assignment statement?

```ballerina
function test() {
    record {|
        int i;
    |} rec = {
        i: 1
    };

    rec.i >>= genInt();
}
```

```ballerina
var arr = [[12, 13, 11], [5, 11]];

function test() {
    foreach var arrItem in arr {
        arrItem[0] /= getInt(arrItem[1]);
    }
}
```

File: language-server/modules/langserver-core/src/test/resources/codeaction/create-function/source/createUndefinedFunctionInCompoundStatement3.bal [hunk 1]
```diff
@@ -0,0 +1,4 @@
+function test() {
+    string s = "";
+    s += concat(i);
+}
```
Comment (SasinduDilshara) [inline language-server/modules/langserver-core/src/test/resources/codeaction/create-function/source/createUndefinedFunctionInCompoundStatement3.bal:4]: Shall we add a test case in the module level as well?

**Review by nipunayf — APPROVED** [review]
Since the change only reuses existing functions—the `typeOf` API and the `checkAndSetTypeResult—the existing tests are sufficient. Fundamentally, we only need to test the compound statement path, and we can assume the reused functions have been tested correctly. Feel free to add the above test, but the changes LGTM.

File: language-server/modules/langserver-core/src/test/resources/codeaction/create-function/source/createUndefinedFunctionInCompoundStatement3.bal (hunk 1 above)
Comment (KavinduZoysa) [reply language-server/modules/langserver-core/src/test/resources/codeaction/create-function/source/createUndefinedFunctionInCompoundStatement3.bal:4]: Added

File: language-server/modules/langserver-core/src/test/resources/codeaction/create-function/config/createUndefinedFunctionInCompoundStatement4.json [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+      "line": 6,
+      "character": 9
+    },
+    "end": {
+      "line": 6,
+      "character": 17
+    }
+  },
+  "source": "testproject/modules/module2/module2.bal",
+  "expected": []
```
Comment (mindula) [inline language-server/modules/langserver-core/src/test/resources/codeaction/create-function/config/createUndefinedFunctionInCompoundStatement4.json:13]: We should suggest the code action for this scenario.

File: language-server/modules/langserver-core/src/test/resources/codeaction/create-function/source/createUndefinedFunctionInCompoundStatement3.bal (hunk 1 above)
Comment (KavinduZoysa) [reply language-server/modules/langserver-core/src/test/resources/codeaction/create-function/source/createUndefinedFunctionInCompoundStatement3.bal:4]: @SasinduDilshara, If we have a function invocation like `mod1:concat(i)`, we do not suggest creating function CA in `mod1`.

File: language-server/modules/langserver-core/src/test/resources/codeaction/create-function/config/createUndefinedFunctionInCompoundStatement4.json (hunk 2 above)
Comment (KavinduZoysa) [reply language-server/modules/langserver-core/src/test/resources/codeaction/create-function/config/createUndefinedFunctionInCompoundStatement4.json:13]: Remove this test

**Review by mindula — APPROVED** [review]
LGTM

**Pull request merged**